src/test/fft.h

```diff
@@ -199,6 +223,209 @@ namespace Algorithms {
+                    vec1[k * L + j] = vec1[k * L + j] + tau; //FAILS
+#else //CALCULATE_WITH_COMPLEX_DATATYPE
+                    tau = complex(
+                            w.real()*vec1[k * L + j + L_s].real() -
+                            w.imag()*vec1[k * L + j + L_s].imag(),
+                            w.imag()*vec1[k * L + j + L_s].real() +
+                            w.real()*vec1[k * L + j + L_s].imag());
+                    vec1[k * L + j + L_s] = complex(
+                            vec1[k * L + j].real() - tau.real(),
+                            vec1[k * L + j].imag() - tau.imag());
+                    vec1[k * L + j] = complex(
+                            vec1[k * L + j].real() + tau.real(),
+                            vec1[k * L + j].imag() + tau.imag());
+#endif //CALCULATE_WITH_COMPLEX_DATATYPE
+                }
+            }
+        }
+
+        return;
+    }
+
+    // Cooley-Tukey type in-place FFT
+    template<class T>
+    static void ct_core(std::vector<std::complex<T>> &vec1,
+                 std::vector<std::complex<T>> &twiddles,
+                 int direction_flag) {
+        unsigned int N = (unsigned int) vec1.size();
+        unsigned int S = (unsigned int) (std::log10((double) N) /
+                                         std::log10((double) 2));
+        unsigned int stride = ((unsigned int) twiddles.size()) * 2 / N;
+
+        unsigned int L, r, L_s, r_s;
+        std::complex<T> w, tau;
+
+        reorder(vec1);
+
+        for (unsigned int q = 1; q <= S; ++q) {
+            L = (unsigned int) 1 << q;
+            L_s = (unsigned int) 1 << (q - 1);
+            r = N >> q;
+            r_s = N >> (q - 1);
+
+            for (unsigned int j = 0; j < L_s; ++j) {
+                w = (direction_flag < 0)
+                    ? twiddles[j * r * stride]
+                    : std::conj(twiddles[j * r * stride]);
+                for (unsigned int k = 0; k < r; ++k) {
+                    tau = w * vec1[k * L + j + L_s];
+                    vec1[k * L + j + L_s] = vec1[k * L + j] - tau;
+                    vec1[k * L + j] = vec1[k * L + j] + tau;
+                }
+            }
+        }
+
+        return;
+    }
+
+    template<class T>
+    static T find_max(std::vector<std::complex<T>> &in) {
+        unsigned int N = (unsigned int) in.size();
+        T max = 0;
+
+        for (unsigned int i = 0; i<N;++i){
+            if (abs(in[i].real()) > max) {
+                max = abs(in[i].real());
+            }
+            if (abs(in[i].imag()) > max) {
+                max = abs(in[i].imag());
+            }
+        }
+        return max;
+    }
+
+    template<class T>
+    static int normalize(std::vector<std::complex<T>> &in, int adjustment) {
+        T max = find_max(in);
+        int norm = leading_bits(max) - max.integer_digits - adjustment;
+        if (norm < 0)  {
+            for (unsigned int i = 0;i<(unsigned int)in.size(); ++i) {
```

Comment:
A similar trick can be done here with `std::transform`.  In general, I recommend watching a Sean Parent talk about removing `for` loops.